File: problems/17/solution.ts
// Accepted: 70ms (5.71%), 51.27MB (64.07%)

function letterCombinations(digits: string): string[] {
    if (digits.length == 0) {
        return [];
    }

    const digitToLetters = {
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxzy"
    };

    const combs: string[] = [];

    function generateCombinations(i: number, collected: string[]): void {
        if (i >= digits.length) {
            const comb = collected.join("");
            combs.push(comb);
            return;
        }

        const digit = digits[i];
        const letters = digitToLetters[digit];
        for (const l of letters) {
            collected.push(l);
            generateCombinations(i+1, collected);
            collected.pop();
        }
    }

    generateCombinations(0, []);

    return combs;
};

File: problems/17/solution.cpp
// Accepted: 0ms (100.00%), 8.59MB (16.58%)

class Solution {
private:
    unordered_map<char, string> digitToLetters {
        { '2', "abc" },
        { '3', "def" },
        { '4', "ghi" },
        { '5', "jkl" },
        { '6', "mno" },
        { '7', "pqrs" },
        { '8', "tuv" },
        { '9', "wxyz" }
    };

    string digits;
    vector<char> collected;
    vector<string> combs;

    void generateCombinations(int i) {
        if (i >= digits.size()) {
            string comb(collected.begin(), collected.end());
            combs.push_back(comb);
            return;
        }

        char digit = digits[i];
        string letters = digitToLetters[digit];
        for (char l : letters) {
            collected.push_back(l);
            generateCombinations(i+1);
            collected.pop_back();
        }
    }

public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) {
            return {};
        }

        this->digits = digits;
        this->collected = {};
        this->combs = {};

        generateCombinations(0);

        return combs;
    }
};
